// next-video/dist/cjs/webpack-loader.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var webpack_loader_exports = {};
__export(webpack_loader_exports, {
  default: () => loader,
  raw: () => raw
});
module.exports = __toCommonJS(webpack_loader_exports);
var import_node_path = __toESM(require("node:path"), 1);
var import_promises = require("node:fs/promises");
var import_node_process = require("node:process");
const raw = true;
async function loader(source) {
  const assetPath = import_node_path.default.resolve(getAssetConfigPath(this.resourcePath));
  this.addDependency(assetPath);
  let asset;
  try {
    asset = await (0, import_promises.readFile)(assetPath, "utf-8");
  } catch {
    const originalFilePath = isRemote(this.resourcePath) ? this.resourcePath : import_node_path.default.relative(process.cwd(), this.resourcePath);
    asset = JSON.stringify({
      status: "sourced",
      originalFilePath
    });
    await (0, import_promises.writeFile)(assetPath, asset, { flag: "wx" });
  }
  return `export default ${asset}`;
}
function getAssetConfigPath(filePath) {
  if (isRemote(filePath)) {
    const VIDEOS_DIR = JSON.parse(import_node_process.env["__NEXT_VIDEO_OPTS"] ?? "{}").folder;
    if (!VIDEOS_DIR)
      throw new Error("Missing video `folder` config.");
    return `${VIDEOS_DIR}/${toSafePath(filePath)}.json`;
  }
  return `${filePath}.json`;
}
function isRemote(filePath) {
  return /^https?:\/\//.test(filePath);
}
function toSafePath(str) {
  return str.replace(/^[^a-zA-Z0-9]+|[^a-zA-Z0-9]+$/g, "").replace(/[^a-zA-Z0-9._-]+/g, "_");
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  raw
});
