// next-video/dist/cjs/cli/init.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc2) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc2 = __getOwnPropDesc(from, key)) || desc2.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var init_exports = {};
__export(init_exports, {
  builder: () => builder,
  command: () => command,
  desc: () => desc,
  handler: () => handler
});
module.exports = __toCommonJS(init_exports);
var import_prompts = require("@inquirer/prompts");
var import_chalk = __toESM(require("chalk"), 1);
var import_node_os = __toESM(require("node:os"), 1);
var import_node_child_process = require("node:child_process");
var import_promises = require("node:fs/promises");
var import_node_path = __toESM(require("node:path"), 1);
var import_logger = __toESM(require("../logger.js"), 1);
var import_json_configs = require("./lib/json-configs.js");
var import_next_config = __toESM(require("./lib/next-config.js"), 1);
const TYPES_FILE_CONTENTS = `/// <reference types="next-video/video-types/global" />
`;
const DEFAULT_DIR = "videos";
const DEV_SCRIPT = "& npx next-video sync -w";
const gitIgnoreContents = (videosDir) => `
# next-video
${videosDir}/*
!${videosDir}/*.json
!${videosDir}/*.js
!${videosDir}/*.ts
public/_next-video
`;
async function preInitCheck(dir) {
  try {
    await (0, import_promises.stat)(dir);
    return false;
  } catch (err) {
    if (err.code === "ENOENT") {
      return true;
    }
    throw err;
  }
}
async function checkVersionManager() {
  try {
    await (0, import_promises.access)("package-lock.json");
    return "npm";
  } catch {
  }
  try {
    await (0, import_promises.access)("yarn.lock");
    return "yarn";
  } catch {
  }
  try {
    await (0, import_promises.access)("pnpm-lock.yaml");
    return "pnpm";
  } catch {
  }
  return void 0;
}
function execPromise(command2) {
  return new Promise((resolve, reject) => {
    (0, import_node_child_process.exec)(command2, (err, stdout, stderr) => {
      if (err) {
        return reject(err);
      }
      return resolve(stdout);
    });
  });
}
async function createVideoDir(dir) {
  const fullPath = import_node_path.default.join(process.cwd(), dir);
  await (0, import_promises.mkdir)(fullPath, { recursive: true });
  await (0, import_promises.appendFile)(".gitignore", gitIgnoreContents(dir));
  return;
}
async function createTSFile(filePath) {
  await (0, import_promises.writeFile)(filePath, TYPES_FILE_CONTENTS);
  return;
}
async function updateTSConfigFile(tsConfigPath) {
  const configContents = await (0, import_promises.readFile)(tsConfigPath, "utf-8");
  const updatedContents = (0, import_json_configs.updateTSConfigFileContent)(configContents);
  return (0, import_promises.writeFile)(tsConfigPath, updatedContents);
}
const command = "init [dir]";
const desc = "Initializes next-video in a project.";
function builder(yargs) {
  return yargs.options({
    dir: {
      alias: "d",
      describe: "The directory you want to initialize next-video with.",
      type: "string"
    },
    force: {
      alias: "f",
      describe: "Continue with initialization even if the chosen directory already exists.",
      type: "boolean",
      default: false
    },
    typescript: {
      alias: "ts",
      describe: "Initialize next-video for use with TypeScript.",
      type: "boolean",
      default: false
    },
    tsconfig: {
      describe: "Automatically update your tsconfig.json file to include the next-video types.",
      type: "boolean",
      default: false
    },
    devscript: {
      describe: `Automatically update your package.json to add the watch command to your dev script.`,
      type: "boolean",
      default: false
    }
  });
}
async function handler(argv) {
  let baseDir = argv.dir;
  let packageInstalled = false;
  let ts = argv.typescript;
  let updateTsConfig = argv.tsconfig;
  let updateDevScript = argv.devscript;
  let changes = [];
  try {
    packageInstalled = await (0, import_json_configs.checkPackageJsonForNextVideo)("./package.json");
  } catch (err) {
    if (err.code === "ENOENT") {
      import_logger.default.error(
        `Failed to find/read a local package.json. Double check that you're running this from the root of your project.`
      );
      return;
    }
    console.log(err);
  }
  if (!packageInstalled) {
    const install = await (0, import_prompts.confirm)({
      message: `It doesn't look like ${import_chalk.default.magenta.bold(
        "next-video"
      )} is installed in this project. Would you like to install it now?`,
      default: true
    });
    if (install) {
      const manager = await checkVersionManager();
      if (!manager) {
        import_logger.default.error("Failed to detect a package manager. Please install next-video manually and re-run this command.");
        import_logger.default.info("For example, in NPM: npm install --save-dev next-video");
        return;
      }
      import_logger.default.info("Detected package manager:", manager);
      import_logger.default.info("Installing next-video...");
      try {
        if (manager === "npm") {
          await execPromise("npm install next-video");
        } else if (manager === "yarn") {
          await execPromise("yarn add next-video");
        } else if (manager === "pnpm") {
          await execPromise("pnpm add next-video");
        }
        import_logger.default.info("Successfully installed next-video!");
      } catch (err) {
        import_logger.default.error("Failed to install next-video:", err);
      }
    } else {
      import_logger.default.info("Make sure to add next-video to your package.json manually");
    }
  }
  if (!baseDir) {
    baseDir = await (0, import_prompts.input)({ message: "What directory should next-video use for video files?", default: DEFAULT_DIR });
  }
  const shouldContinue = await preInitCheck(baseDir);
  if (!argv.force && !shouldContinue) {
    import_logger.default.warning("Directory already exists:", baseDir);
    import_logger.default.info("If you'd like to proceed anyway, re-run with --force");
    return;
  }
  await createVideoDir(baseDir);
  changes.push([import_logger.default.add, `Created ${baseDir} directory.`]);
  if (!ts) {
    ts = await (0, import_prompts.confirm)({ message: "Is this a TypeScript project?", default: true });
  }
  if (ts) {
    await createTSFile(import_node_path.default.join(process.cwd(), "video.d.ts"));
    changes.push([import_logger.default.add, `Created video.d.ts.`]);
  }
  if (ts && !updateTsConfig) {
    updateTsConfig = await (0, import_prompts.confirm)({ message: "Update tsconfig.json to include next-video types?", default: true });
  }
  if (updateTsConfig) {
    try {
      await updateTSConfigFile(import_node_path.default.join(process.cwd(), "tsconfig.json"));
      changes.push([import_logger.default.add, `Updated tsconfig.json to include next-video types.`]);
    } catch (err) {
      changes.push([import_logger.default.error, 'Failed to update tsconfig.json, please add "video.d.ts" to the include array']);
    }
  } else if (ts) {
    changes.push([import_logger.default.info, `Add ${import_chalk.default.underline("video.d.ts")} to the includes array in tsconfig.json.`]);
  }
  const cmd = await isCmd();
  if (!cmd && !updateDevScript) {
    updateDevScript = await (0, import_prompts.confirm)({
      message: `Update package.json to add the watch command to your dev script?`,
      default: true
    });
  }
  if (!cmd && updateDevScript) {
    try {
      const devScript = (await execPromise(`npm pkg get scripts.dev`))?.trim().slice(1, -1);
      if (devScript && !devScript.includes(DEV_SCRIPT)) {
        await execPromise(`npm pkg set scripts.dev='${devScript} ${DEV_SCRIPT}'`);
      }
      changes.push([import_logger.default.add, `Updated package.json to add the watch command to your dev script.`]);
    } catch (err) {
      changes.push([import_logger.default.error, `Failed to update package.json, please add "${DEV_SCRIPT}" to your dev script.`]);
    }
  }
  try {
    const update = await (0, import_next_config.default)("./", { folder: baseDir });
    if (update) {
      changes.push([import_logger.default.add, `Updated ${update.configPath} to include next-video.`]);
    }
  } catch (e) {
    if (e.error === "not_found") {
      changes.push([
        import_logger.default.error,
        "No next.config.js or next.config.mjs file found. Please add next-video to your config manually."
      ]);
    } else if (e.error === "already_added") {
      changes.push([import_logger.default.info, "It seems like next-video is already added to your Next Config"]);
    } else {
      changes.push([import_logger.default.error, "Failed to update next.config.js, please add next-video to your config manually."]);
    }
  }
  import_logger.default.success(`${import_chalk.default.magenta.bold("next-video")} initialized!`);
  changes.forEach(([loggerFn, change]) => loggerFn(change));
  import_logger.default.info("");
  import_logger.default.info(`NEXT STEP: Set up remote storage`);
  import_logger.default.info(import_chalk.default.magenta.bold("https://next-video.dev/docs#remote-storage-and-optimization"));
  import_logger.default.info("");
}
async function isCmd() {
  if (import_node_os.default.platform() === "win32") {
    try {
      await execPromise(`ls`);
    } catch (err) {
      return true;
    }
  }
  return false;
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  builder,
  command,
  desc,
  handler
});
