// next-video/dist/cjs/providers/mux.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var mux_exports = {};
__export(mux_exports, {
  createThumbHash: () => createThumbHash,
  uploadLocalFile: () => uploadLocalFile,
  uploadRequestedFile: () => uploadRequestedFile
});
module.exports = __toCommonJS(mux_exports);
var import_promises = __toESM(require("node:fs/promises"), 1);
var import_node_path = __toESM(require("node:path"), 1);
var import_chalk = __toESM(require("chalk"), 1);
var import_mux_node = __toESM(require("@mux/mux-node"), 1);
var import_undici = require("undici");
var import_sharp = __toESM(require("sharp"), 1);
var import_assets = require("../assets.js");
var import_logger = __toESM(require("../logger.js"), 1);
let mux;
function initMux() {
  mux = new import_mux_node.default();
}
function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}
async function pollForAssetReady(filePath, asset) {
  if (!asset.externalIds?.assetId) {
    import_logger.default.error("No assetId provided for asset.");
    console.error(asset);
    return;
  }
  initMux();
  const assetId = asset.externalIds?.assetId;
  const muxAsset = await mux.video.assets.retrieve(assetId);
  const playbackId = muxAsset.playback_ids?.[0].id;
  let updatedAsset = asset;
  if (asset.externalIds?.playbackId !== playbackId) {
    updatedAsset = await (0, import_assets.updateAsset)(filePath, {
      externalIds: {
        playbackId
      }
    });
  }
  if (muxAsset.status === "errored") {
    import_logger.default.error(import_logger.default.label("Asset errored:"), filePath);
    import_logger.default.space(import_chalk.default.gray(">"), import_logger.default.label("Mux Asset ID:"), assetId);
    return (0, import_assets.updateAsset)(filePath, {
      status: "error",
      error: muxAsset.errors
    });
  }
  if (muxAsset.status === "ready") {
    let blurDataURL;
    try {
      blurDataURL = await createThumbHash(`https://image.mux.com/${playbackId}/thumbnail.png?width=100&height=100`);
    } catch (e) {
      import_logger.default.error("Error creating a thumbnail hash.");
    }
    import_logger.default.success(import_logger.default.label("Asset is ready:"), filePath);
    import_logger.default.space(import_chalk.default.gray(">"), import_logger.default.label("Playback ID:"), playbackId);
    return (0, import_assets.updateAsset)(filePath, {
      status: "ready",
      externalIds: {
        playbackId
      },
      blurDataURL
    });
  } else {
    await sleep(1e3);
    return pollForAssetReady(filePath, updatedAsset);
  }
}
async function pollForUploadAsset(filePath, asset) {
  if (!asset.externalIds?.uploadId) {
    import_logger.default.error("No uploadId provided for asset.");
    console.error(asset);
    return;
  }
  initMux();
  const uploadId = asset.externalIds?.uploadId;
  const muxUpload = await mux.video.uploads.retrieve(uploadId);
  if (muxUpload.asset_id) {
    import_logger.default.info(import_logger.default.label("Asset is processing:"), filePath);
    import_logger.default.space(import_chalk.default.gray(">"), import_logger.default.label("Mux Asset ID:"), muxUpload.asset_id);
    const processingAsset = await (0, import_assets.updateAsset)(filePath, {
      status: "processing",
      externalIds: {
        assetId: muxUpload.asset_id
      }
    });
    return pollForAssetReady(filePath, processingAsset);
  } else {
    await sleep(1e3);
    return pollForUploadAsset(filePath, asset);
  }
}
async function uploadLocalFile(asset) {
  if (!asset.originalFilePath) {
    import_logger.default.error("No filePath provided for asset.");
    console.error(asset);
    return;
  }
  initMux();
  if (asset.status === "ready") {
    return;
  } else if (asset.status === "processing") {
    import_logger.default.info(import_logger.default.label("Asset is already processing. Polling for completion:"), asset.originalFilePath);
    return pollForAssetReady(asset.originalFilePath, asset);
  } else if (asset.status === "uploading") {
    import_logger.default.info(import_logger.default.label("Resuming upload:"), asset.originalFilePath);
  }
  const src = asset.originalFilePath;
  if (src && /^https?:\/\//.test(src)) {
    return uploadRequestedFile(asset);
  }
  let upload;
  try {
    upload = await mux.video.uploads.create({
      cors_origin: "*",
      // @ts-ignore
      new_asset_settings: {
        playback_policy: ["public"]
      }
    });
  } catch (e) {
    import_logger.default.error("Error creating a Mux Direct Upload");
    console.error(e);
    return;
  }
  await (0, import_assets.updateAsset)(src, {
    status: "uploading",
    externalIds: {
      uploadId: upload.id
      // more typecasting while we use the beta mux sdk
    }
  });
  const filePath = import_node_path.default.join(src);
  const fileDescriptor = await import_promises.default.open(filePath);
  const fileStats = await fileDescriptor.stat();
  const stream = fileDescriptor.createReadStream();
  import_logger.default.info(import_logger.default.label("Uploading file:"), `${filePath} (${fileStats.size} bytes)`);
  try {
    await (0, import_undici.fetch)(upload.url, {
      method: "PUT",
      // @ts-ignore
      body: stream,
      duplex: "half"
    });
    stream.close();
    await fileDescriptor.close();
  } catch (e) {
    import_logger.default.error("Error uploading to the Mux upload URL");
    console.error(e);
    return;
  }
  import_logger.default.success(import_logger.default.label("File uploaded:"), `${filePath} (${fileStats.size} bytes)`);
  const processingAsset = await (0, import_assets.updateAsset)(src, {
    status: "processing"
  });
  return pollForUploadAsset(src, processingAsset);
}
async function uploadRequestedFile(asset) {
  if (!asset.originalFilePath) {
    import_logger.default.error("No URL provided for asset.");
    console.error(asset);
    return;
  }
  initMux();
  if (asset.status === "ready") {
    return;
  } else if (asset.status === "processing") {
    import_logger.default.info(import_logger.default.label("Asset is already processing. Polling for completion:"), asset.originalFilePath);
    return pollForAssetReady(asset.originalFilePath, asset);
  }
  const src = asset.originalFilePath;
  const assetObj = await mux.video.assets.create({
    // @ts-ignore
    input: [{
      url: asset.originalFilePath
    }],
    playback_policy: ["public"]
  });
  import_logger.default.info(import_logger.default.label("Asset is processing:"), src);
  import_logger.default.space(import_chalk.default.gray(">"), import_logger.default.label("Mux Asset ID:"), assetObj.id);
  const processingAsset = await (0, import_assets.updateAsset)(src, {
    status: "processing",
    externalIds: {
      assetId: assetObj.id
    }
  });
  return pollForAssetReady(src, processingAsset);
}
async function createThumbHash(imgUrl) {
  const response = await (0, import_undici.fetch)(imgUrl);
  const buffer = await response.arrayBuffer();
  const { data, info } = await (0, import_sharp.default)(buffer).raw().ensureAlpha().toBuffer({ resolveWithObject: true });
  const { rgbaToThumbHash, thumbHashToDataURL } = await import("thumbhash");
  const hash = rgbaToThumbHash(info.width, info.height, data);
  return thumbHashToDataURL(hash);
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  createThumbHash,
  uploadLocalFile,
  uploadRequestedFile
});
