// next-video/dist/cjs/providers/vercel-blob.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var vercel_blob_exports = {};
__export(vercel_blob_exports, {
  config: () => config,
  uploadLocalFile: () => uploadLocalFile,
  uploadRequestedFile: () => uploadRequestedFile
});
module.exports = __toCommonJS(vercel_blob_exports);
var import_promises = __toESM(require("node:fs/promises"), 1);
var import_node_path = __toESM(require("node:path"), 1);
var import_undici = require("undici");
var import_blob = require("@vercel/blob");
var import_chalk = __toESM(require("chalk"), 1);
var import_assets = require("../assets.js");
var import_logger = __toESM(require("../logger.js"), 1);
const config = {
  runtime: "edge"
};
async function uploadLocalFile(asset) {
  if (!asset.originalFilePath) {
    import_logger.default.error("No filePath provided for asset.");
    console.error(asset);
    return;
  }
  if (asset.status === "ready") {
    return;
  } else if (asset.status === "uploading") {
    import_logger.default.info(import_logger.default.label("Resuming upload:"), asset.originalFilePath);
  }
  const src = asset.originalFilePath;
  await (0, import_assets.updateAsset)(src, {
    status: "uploading"
  });
  if (src && /^https?:\/\//.test(src)) {
    return uploadRequestedFile(asset);
  }
  const filePath = import_node_path.default.join(src);
  const fileDescriptor = await import_promises.default.open(filePath);
  const fileStats = await fileDescriptor.stat();
  const stream = fileDescriptor.createReadStream();
  import_logger.default.info(import_logger.default.label("Uploading file:"), `${filePath} (${fileStats.size} bytes)`);
  let blob;
  try {
    blob = await (0, import_blob.put)(asset.originalFilePath, stream, { access: "public" });
    stream.close();
    await fileDescriptor.close();
  } catch (e) {
    import_logger.default.error("Error uploading to the Vercel Blob store");
    console.error(e);
    return;
  }
  import_logger.default.success(import_logger.default.label("File uploaded:"), `${filePath} (${fileStats.size} bytes)`);
  import_logger.default.space(import_chalk.default.gray(">"), import_logger.default.label("URL:"), blob.url);
  return (0, import_assets.updateAsset)(src, {
    status: "ready",
    externalIds: {
      url: blob.url
    }
  });
}
async function uploadRequestedFile(asset) {
  if (!asset.originalFilePath) {
    import_logger.default.error("No URL provided for asset.");
    console.error(asset);
    return;
  }
  if (asset.status === "ready") {
    return;
  }
  const src = asset.originalFilePath;
  const fileName = import_node_path.default.basename(src);
  const stream = (await (0, import_undici.fetch)(src)).body;
  if (!stream) {
    import_logger.default.error("Error fetching the requested file:", src);
    return;
  }
  import_logger.default.info(import_logger.default.label("Uploading file:"), src);
  let blob;
  try {
    blob = await (0, import_blob.put)(fileName, stream, { access: "public" });
  } catch (e) {
    import_logger.default.error("Error uploading to the Vercel Blob store");
    console.error(e);
    return;
  }
  import_logger.default.success(import_logger.default.label("File uploaded:"), src);
  import_logger.default.space(import_chalk.default.gray(">"), import_logger.default.label("URL:"), blob.url);
  return (0, import_assets.updateAsset)(src, {
    status: "ready",
    externalIds: {
      url: blob.url
    }
  });
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  config,
  uploadLocalFile,
  uploadRequestedFile
});
